File: VerbumKit/Sources/Features/Explore/EntityListFeature.swift
import Clients
import ComposableArchitecture
import Models

/// Every entity of one kind — People, Places, Themes, Events (spec §7).
@Reducer
public struct EntityListFeature {
    @ObservableState
    public struct State: Equatable {
        public let type: BibleEntityType
        public var entities: [BibleEntity] = []
        public var isLoading = false

        public init(type: BibleEntityType) {
            self.type = type
        }
    }

    public enum Action: Equatable {
        case task
        case entitiesResponse([BibleEntity])
        case entityTapped(BibleEntity)
        case delegate(Delegate)

        @CasePathable
        public enum Delegate: Equatable {
            case openEntity(BibleEntity)
        }
    }

    @Dependency(\.graphClient) var graphClient

    public init() {}

    public var body: some ReducerOf<Self> {
        Reduce { state, action in
            switch action {
            case .task:
                state.isLoading = true
                let type = state.type
                return .run { [graphClient] send in
                    await send(.entitiesResponse((try? await graphClient.entities(type: type)) ?? []))
                }

            case .entitiesResponse(let entities):
                state.isLoading = false
                state.entities = entities
                return .none

            case .entityTapped(let entity):
                return .send(.delegate(.openEntity(entity)))

            case .delegate:
                return .none
            }
        }
    }
}

File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/EntityListFeature.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.clients.GraphClient
import com.nexussoft.verbum.common.arch.Effect
import com.nexussoft.verbum.common.arch.Reducer
import com.nexussoft.verbum.common.arch.only
import com.nexussoft.verbum.common.arch.runEffect
import com.nexussoft.verbum.common.arch.with
import com.nexussoft.verbum.models.BibleEntity
import com.nexussoft.verbum.models.BibleEntityType

/** Every entity of one kind — People, Places, Themes, Events (§7). Twin of iOS `EntityListFeature`. */
object EntityListFeature {
    data class State(val type: BibleEntityType, val entities: List<BibleEntity> = emptyList(), val isLoading: Boolean = false)

    sealed interface Action {
        data object Started : Action
        data class EntitiesLoaded(val entities: List<BibleEntity>) : Action
        data class EntityTapped(val entity: BibleEntity) : Action
        data class Delegate(val delegate: DelegateAction) : Action
    }

    sealed interface DelegateAction {
        data class OpenEntity(val entity: BibleEntity) : DelegateAction
    }

    fun reducer(graphClient: GraphClient): Reducer<State, Action> = Reducer { state, action ->
        when (action) {
            Action.Started -> state.copy(isLoading = true).with(runEffect { send -> send(Action.EntitiesLoaded(runCatching { graphClient.entities(state.type) }.getOrDefault(emptyList()))) })
            is Action.EntitiesLoaded -> state.copy(isLoading = false, entities = action.entities).only()
            is Action.EntityTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.OpenEntity(action.entity))))
            is Action.Delegate -> state.only()
        }
    }
}
